### ISSUE TITLE
SQL Server: Support hierarchyid

### ISSUE THREAD

**Issue by mojtabakaviani** [dotnet/efcore#365]
We are using this issue to cover specifically end-to-end support for hierarchyid. SQL Server UDTs are covered by https://github.com/aspnet/EntityFrameworkCore/issues/12796, and spatial is covered by https://github.com/aspnet/EntityFrameworkCore/issues/1100. 

As noted in https://github.com/aspnet/EntityFrameworkCore/issues/365#issuecomment-338293994, we already support mapping a property of this type where the type is available. 

Here is what is still missing before we can say we support hierarchyid:
- Currently the type is only available in .NET Framework (mostly because Microsoft.SqlServer.Types hasn't been ported)
- Presumably we should support usage of members of the type, functions, etc. in queries

## Original issue
sql server special data types add in EF5 and later but hierarchyid and user defined types not implemented yet.please work on hierarchyid and user defined types that use in real website and enterprise.

Note by @rowanmiller: Also relevant - [[C# Feature Request] Hierarchy Data Type #16193](https://github.com/dotnet/roslyn/issues/16193)

**Comment by romcoudray** [comment]
Have you someting planned yet to implement hierarchyid in EF7? Looking forward for your thoughts on this topic, thanks.

**Comment by vRITHNER** [comment]
Hey Rowan, just to add my 2 ct, hope you'll support HierarchyId asap because it's really a great way to manage hierarchy and other manual ways demand so much effort in code. 
Also, it could be great to support natively the GROUP BY xxxx WITH ROLLUP to allow to calculate par ex. Totals per group/subgroup/etc... witch make a huge save code effort + offcourse performance improvement 
Txs
-Vince

**Comment by tomagnew** [comment]
Please add native support for hierarchyID to EF7 so we can avoid work-arounds (like sprocs which defeat the purpose of an ORM) to use this nice feature of Sql Server 2008 and greater. IMHO, hierarchyId is the most efficient way to work with hierarchical data which helps reduce the impedance between normalized table data and nested data.

**Comment by leus** [comment]
Don't know if this is proper protocol, but I'd like hierarchyid gets some love too - too many times in my career I had to create tables with "parent_id" and stupidly complex recursive queries just to draw a tree (and just now I'm doing it again)

**Comment by nicbourgoin** [comment]
Any guidance on where I should start looking for to try and implement this? I definitly need hierarchicalid from Microsoft.SqlServer.Types to be supported.

**Comment by nteague22** [comment]
HierarchyId might be work tackling sooner than later, if for no other reason than doing the Reference<>() definitions and Collection<>() definitions in the IModel.  Parenting chains within a table could greatly make steps to reduce the separation I have had with using the outstanding features of the system, while juggling the inability to do extensive business and enterprise processing

**Comment by weitzhandler** [comment]
Any news?
See [this](https://github.com/aspnet/EntityFramework/issues/3241) issue too.

**Comment by activebiz** [comment]
Any update on this ?

**Comment by ravetroll** [comment]
I'd like the upvote this too. #365 was opened in 2014 and its now 2017.  Hierarchical models are commonplace and the SQL hierarchy type does greatly improve performance on deep tree requests over he iterative parent/child approach.

**Comment by activebiz** [comment]
[Rant warning :)] I think the whole EF stuff is too much to bring small gain on writing efficient code. I have been using EF since 4x days but it always have issues with either lack of support for certain types, performance etc. Its great ORM (dont get me wrong) and might be ok for certain projects but I think that no Full-ORM is a match for good old ADO.NET. I have recently backed out from EF for reason mentioned above and ofcourse the performance.  I think using full flege ORM will always have issues becouse its in some sense making decistions for you. For e.g. creating queries etc.  If I have to see SQL Profiler after writing my Linq-EF queries everytime then its defeating the purpose of writing efficient code fast.

I think EF should offer another version which is Mirco ORM such as Dapper.  This will elimiates lot of these issues with not supproting this type and that type and writing/profiling queries etc. 

@ravetroll . I would look at Dapper or PetaPoco as another alternative to EF. Dapper does support HierarcyId if you are using Full .NET framework. (its sort of .NET core limitations of not supporing hierarcyId). 

Sorry to be so much negative about this but I have waited for EF to bring HierarchyId for so long. (even 6x doesnt support it out of box , you have to use external package for that). Just shaing my pain.

**Comment by weitzhandler** [comment]
@activebiz 
I've been upgrading along with EF until it came to EFCore. And I'll upgrade to core as soon as complex-types and TPT is implemented.
I do agree with you that it could be nice to see some portions of EF are extract to independent libs, such as change-tracking etc.
But I do love the EF, and it only keeps on getting better.

**Comment by joshcomley** [comment]
Does anybody have any examples of HierarchyId on EF Core?

**Comment by activebiz** [comment]
@joshcomley It doesnt support simply becouse dotnet core dont support. So dont hold your breath! If you want you can use EF6 and there is a nuget package which you have to use.

https://www.nuget.org/packages/EntityFrameworkWithHierarchyId/

**Comment by joshcomley** [comment]
@activebiz isn't is a case of building in the support into EF Core? I've checked out that package and it *should* be possible to port the feature to EF Core, with the right know-how

**Comment by activebiz** [comment]
@joshcomley Its possible if dotnet core support HierarchyId which it doesnt at the moment.

**Comment by ajcvickers** [comment]
Just an FYI to this thread that mapping of HierarchyId (as well as SqlGeography and SqlGeometry) is now supported (and will be in 2.1), but with some significant limitations:

* There is no attempt by EF Core to use the semantics of the HierarchyId for mapping or anything else. It is just an opaque object as far as EF is concerned.
* It only works when running on .NET Framework. It will not work on .NET Core, since SqlClient for .NET Core does not support these types.
* Query support is limited:
  * LINQ queries that don't use the semantics of the SQL types should work.
  * Queries that need to use the type semantics (e.g. calling functions on the HierarchyId object) can be done with FromSql, but not LINQ
  * See issue #10108 for some test code and issue #10109 for some ideas to make LINQ work better

**Comment by olmobrutall** [comment]
I've just made a [pull request](https://github.com/dotMorten/Microsoft.SqlServer.Types/pull/6 ) adding support for `SqlHierarchyId` in [dotMorten's ](https://github.com/dotMorten/Microsoft.SqlServer.Types). 

When merged, the repository will have a .Net Standard implementation of all the important types in  `Microsoft.SqlServer.Types`: 
* `SqlHierarchyId` 
* `SqlGeometry`
* `SqlGeography`

Maybe could be an possibility for EntityFrameworkCore and/or SQL Server team?

**Comment by bricelam** [comment]
@olmobrutall It should be trivial to create a package like our [NetTopologySuite ones](https://github.com/aspnet/EntityFrameworkCore/tree/release/2.2/src/EFCore.Sqlite.NTS) that allow these types to be mapped and translated.

One of the reasons we didn't take this approach is because we wanted to enable using the same CLR types (the NetTopologySite spatial types) across multiple providers (e.g. SQL Server, SQLite and PostgreSQL). Encouraging everyone to use the SQL Server types, or only implementing spatial support on SQL Server seemed a bit ...well Microsoft circa 1998. 😉

**Comment by bricelam** [comment]
I looked on NuGet.org for popular type that exposed this functionality on .NET Standard (similar to how NTS enabled us to implement spatial), but couldn't find one. I think we can consider ourselves blocked on this until one emerges or Microsoft.SqlServer.Types is updated to target .NET Standard.

Technically speaking, we (or someone else) *could* create a package that only works on .NET Framework, but this doesn't align with our strategy for EF Core.

**Comment by ErikEJ** [comment]
@dotmorten ?

**Comment by dotMorten** [comment]
I have a SQL hiarchy implementation that matches the SQL types here: 
https://github.com/dotMorten/Microsoft.SqlServer.Types/tree/master/src/Microsoft.SqlServer.Types/SqlHierarchy

**Comment by bricelam** [comment]
Have you published a NuGet package? If so I’ll throw something together during my moonlit hours.

**Comment by dotMorten** [comment]
It was published without consent and because I was procrastinating 😁
I'll have a proper one up with source linking and correctly code signed.

**Comment by bricelam** [comment]
Wow these types are awful. The operator overloads return `SqlBoolean` forcing you to write LINQ like this:
``` csharp
var root = from t in db.Things
           where (bool)(t.HierarchyId == SqlHierarchyId.GetRoot())
           select t;
```
We probably need a more .NET-idiomatic wrapper type to make the EF Core experience acceptable.

**Comment by dotMorten** [comment]
These are an exact match on the sqltypes API. Take it up with Microsoft. It's an API geared towards working with SQL server.

**Comment by dotMorten** [comment]
This is all just how UDT types work in SQL. Not sure what the point of ranting against that is, unless you hope to change how SQL Server works?

**Comment by olmobrutall** [comment]
Exactly

**Comment by dotMorten** [comment]
Fixed several bugs and inconsistencies with the official Sql Types library and uploaded to NuGet: https://www.nuget.org/packages/dotMorten.Microsoft.SqlServer.Types/1.0.0

**Comment by bricelam** [comment]
No ranting here—just remembering why we had to introduce DbGeography in EF6.

I’m also disappointed that fixing this issue isn’t as trivial as I first thought it would be.

**Comment by bricelam** [comment]
I suspect all of the awkwardness stems from the three-valued logic of SQL (True/False/Unknown)

**Comment by dotMorten** [comment]
Now I can't pretend I know much about EF, but I don't understand the big surprise about this. It's not really any different than working with all the built-in SqlTypes. Wouldn't you just do it all the same way?

**Comment by dotMorten** [comment]
FYI OMD of the library here: https://sharpgis.net/omds/dotMorten.Microsoft.SqlServer.Types.html
And here's a list of the API's the official package has, that no .NET Standard implementation is provided for (yet):
https://sharpgis.net/omds/dotMorten.Microsoft.SqlServer.Types-missing.html

**Comment by jmounce** [comment]
I'm having trouble using SqlHierarchyId in an EFCore project. 

Hey, @dotMorten - I've seen your posts and tried your lib. https://github.com/dotMorten/Microsoft.SqlServer.Types
For writes to the DB and DB generation, it seems to work fine. 

But when I read, I get this error:
`FileLoadException: Could not load file or assembly 'Microsoft.SqlServer.Types, Version=10.0.0.0, Culture=neutral, PublicKeyToken=89845dcd8080cc91'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, string codeBase, RuntimeAssembly locationHint, ref StackCrawlMark stackMark, IntPtr pPrivHostBinder, bool throwOnFileNotFound, IntPtr ptrLoadContextBinder)
System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, RuntimeAssembly reqAssembly, ref StackCrawlMark stackMark, IntPtr pPrivHostBinder, bool throwOnFileNotFound, IntPtr ptrLoadContextBinder)
System.Reflection.Assembly.Load(AssemblyName assemblyRef)
System.Data.SqlClient.SqlConnection.ResolveTypeAssembly(AssemblyName asmRef, bool throwOnError)
System.Data.SqlClient.SqlConnection+<>c__DisplayClass128_0.<CheckGetExtendedUDTInfo>b__0(AssemblyName asmRef)
System.TypeNameParser.ResolveAssembly(string asmName, Func<AssemblyName, Assembly> assemblyResolver, bool throwOnError, ref StackCrawlMark stackMark)
System.TypeNameParser.ConstructType(Func<AssemblyName, Assembly> assemblyResolver, Func<Assembly, string, bool, Type> typeResolver, bool throwOnError, bool ignoreCase, ref StackCrawlMark stackMark)
System.TypeNameParser.GetType(string typeName, Func<AssemblyName, Assembly> assemblyResolver, Func<Assembly, string, bool, Type> typeResolver, bool throwOnError, bool ignoreCase, ref StackCrawlMark stackMark)
System.Type.GetType(string typeName, Func<AssemblyName, Assembly> assemblyResolver, Func<Assembly, string, bool, Type> typeResolver, bool throwOnError)
System.Data.SqlClient.SqlConnection.CheckGetExtendedUDTInfo(SqlMetaDataPriv metaData, bool fThrow)
System.Data.SqlClient.SqlDataReader.GetValueFromSqlBufferInternal(SqlBuffer data, _SqlMetaData metaData)
System.Data.SqlClient.SqlDataReader.GetFieldValueFromSqlBufferInternal<T>(SqlBuffer data, _SqlMetaData metaData)
System.Data.SqlClient.SqlDataReader.GetFieldValueInternal<T>(int i)
System.Data.SqlClient.SqlDataReader.GetFieldValue<T>(int i)
lambda_method(Closure , DbDataReader )`

I know that Microsoft.SqlServer.Types doesn't support .NET Standard so your lib looked attractive, but why is the runtime looking for a specific version of this assembly?

I see in the EF6 code that it is specific:
https://github.com/aspnet/EntityFramework6/blob/master/src/EntityFramework.SqlServer/SqlTypesAssemblyLoader.cs#L16

but I don't see similar, specific code in EF Core. Any idea what is going on here or how to fix this?
(other than MS porting to standard - which isn't open source...??)

**Comment by olmobrutall** [comment]
Hi @jmounce,

Check https://github.com/dotMorten/Microsoft.SqlServer.Types/blob/master/src/Microsoft.SqlServer.Types.Tests/DBTests.cs

And search for `reader.GetFieldValue<SqlHierarchyId>`, maybe it helps.

**Comment by jmounce** [comment]
I compiled and used your v1.1 but it didn't work for me. Same error. I have to find out if it is just me or not.
Your tests that @olmobrutall  referred me to don't use EF, so that doesn't tell me much.
I will have to modify your tests and see if it also fails there.

**Comment by olmobrutall** [comment]
There is not support for `SqlHierarchyId` in EF Core yet. Whether they will implement it based on the version that I contributed to dotMorten repo, or wait for an official port, is something EF devs have to answer. 

About base ADO.Net support, I'm also getting your exception now ...strange. Looks like `GetFieldValue<T>(ordinal)` is just `(T)GetValue(ordinal)`, so the type resolution is on ADO.Net. I've been bitten many times by Assembly resolution conflicts in `Microsoft.SqlServer.Types` before (in .Net Framework) so I've changed the implementation now to instantiate the object myself (with `Activator.CreateInstance`), cast it to `IBinarySerializable` and read the cell `Stream`. 

 I've solved it like this in my ORM: https://github.com/signumsoftware/framework/commit/1cf75072b5e823528bcddc931e6cf637aa1ce6ea, maybe is useful for a future EF implementation.

**Comment by dotMorten** [comment]
@jmounce I'm now finding that my signature key hack doesn't work on .NET Framework (it works on .NET Core and Mono). I think @bricelam has a way to hack it more, but I couldn't get it to work.

**Comment by jmounce** [comment]
@dotMorten I'm running on .NET Core 2.1. It's strange. I don't always get this error for all projects. For example, if I add EF support to your test project, it works fine. However, for any WebApi Core projects I create, I get the error. It must be a dependency in ASP.NET Core->EF Core. I just can't determine the cause yet.

**Comment by dotMorten** [comment]
Note: Some pretty serious bugs were reported in the SqlTypes nuget over the weekend and I fixed them all and updated to v1.1.0 and unlisted 1.0.0 (none of these were hierarchy specific though, but they could create data corruption for the geometry types)

**Comment by bricelam** [comment]
I started working on something in [bricelam/EFCore.SqlServer.HierarchyId](https://github.com/bricelam/EFCore.SqlServer.HierarchyId), but had to put it aside for while. I think the only part still missing is adding a type mapping similar to [SqlServerGeometryTypeMapping](https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.SqlServer.NTS/Storage/Internal/SqlServerNetTopologySuiteTypeMappingSourcePlugin.cs#L43). Feel free to pick up where I left off.

**Comment by babak-f** [comment]
Until we get some sort of support for hierarchy ID in EF Core, would it be possible to introduce it as a C# property in code-first, and map it to a hierarchyid column in the database? (as a potential workaround)

**Comment by dotMorten** [comment]
@babak-f my library has hierarchy id support : https://github.com/dotMorten/Microsoft.SqlServer.Types/issues/1

**Comment by aljones** [comment]
I just took a first pass at adding a type mapping for HierarchyId based on @bricelam helpful pointer [here](https://github.com/aljones/EFCore.SqlServer.HierarchyId/commit/e683fe011d4e303d60b94846f88ecd1d4f1dff7e). I'm unfamiliar with these APIs so I may have missed something.  Got some tests passing at least.

**Comment by huoyaoyuan** [comment]
@aljones Happy to see some "just work" solutions.
Your implementation works with 2.2 but not 3.0, since `IDbContextOptionsExtensionWithDebugInfo` has changed.
Creating migrations with `dotnet ef` works, but running that migration results in NRE.
Querying data with relationship works.

I think we should wait `Microsoft.Data.SqlClient` to be opened, then use it for further official support.

**Comment by darthkurak** [comment]
Hello. Can someone explain me a bit:
I'm trying to use SqlHierarchyId in my PoC. First, i used Microsoft.SqlServer.Types nuget package to include it. Everything seemed to work, but when i looked at produced query on SQL Profiler for such example:
Where(p => p.HierarchyId.IsDescendantOf(anotherHierarchyId)) that is completle ingored, and i have simple From without Where.
But in memory, items are filtered and i got only descendants. 
I checked the same under dotMorten nuget package. Same behavior.
This is known (that hierarchyId functions are only applied in memory, after getting items from database) or I'm doing something wrong?

**Comment by darthkurak** [comment]
Also, i checked aljones solution and it worked pretty well, beside one issue:
You can't call sqlHierarchy functions on variables - exception is thrown.
Example:
Where(p => variable.IsDescendantOf(p)) (for finding all ancestors) 
Exception:
System.Data.SqlClient.SqlException: 'Cannot call methods on varbinary.'

**Comment by bricelam** [comment]
@darthkurak You need to specify the function mappings:
``` csharp
modelBuilder
    .HasDbFunction(typeof(SqlHierarchyId).GetMethod(nameof(SqlHierarchyId.IsDescendantOf)))
    .HasTranslation(
        e => SqlFunctionExpression.Create(
            e.First(),
            "IsDescendantOf",
            e.Skip(1),
            typeof(bool),
            null)));
```

**Comment by bricelam** [comment]
Wait. @smitpatel Is the first expression the instance? How do you map instance methods?

**Comment by smitpatel** [comment]
Your code snippet is correct. That is how you map instance methods.

**Comment by bricelam** [comment]
> Cannot call methods on varbinary

Can you leverage [this awesome feature](https://github.com/dotnet/corefx/issues/35470) like we did [for spatial](https://github.com/aspnet/EntityFrameworkCore/pull/16097)?

**Comment by darthkurak** [comment]
@bricelam Actually this is something that i tried to do on my own, (on aljones code) but then i hit exception:
"Specified type is not registered on the target server. System.Data.SqlTypes.SqlBytes, System.Data.Common, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a.'"

Tried to fight with it, but with no success. 
Have any ideas and suggestions?

**Comment by aljones** [comment]
I made the suggested changes from @bricelam and other changes to support efcore 3. 

@darthkurak I got the same error before updating to efcore 3. I think ef core 3 is using Microsoft.Data.SqlClient which enables the new feature, but I haven't followed that trail to be sure.

**Comment by bricelam** [comment]
Yes, the feature was added to SqlClient in the 3.0 timeframe

**Comment by bricelam** [comment]
@aljones Do you ship a NuGet package? If so, we should add it to our [list of extensions](https://docs.microsoft.com/en-us/ef/core/extensions/).

**Comment by huoyaoyuan** [comment]
My edited version works for simple scenarios, but breaks for second migration(initial was good).
I want an officially supported version that doesn't have corner case issues.

**Comment by darthkurak** [comment]
Thanks for update. Is there possibility to introduce this change in 2.2? We will not update our project to 3.0 yet, and would be great to have working hierarchyId in EF core. Also I agree that definitely we need official version. Is there any chance for that in near future?

**Comment by aljones** [comment]
I haven't published a nuget package. The project I was going to use this for was delayed. I really only know that a few simple tests pass.  If the project gets rescheduled then I can justify spending more time on it, but not until then.

**Comment by chrisrichards** [comment]
I'm struggling to get migrations to run in EF Core with a HierarchyId column.

I've created a [test project](https://github.com/chrisrichards/EFCore.SqlServer.HierarchyId) that is based on the code from @aljones and @dotMorten but I've replaced the dependency on System.Data.SqlClient with Microsoft.Data.SqlClient.

The QueryTests all pass, but when I attempt to generate migrations from the EFCore.SqlServer.HierarchyId.Test folder it fails as below:

```bash
dotnet ef migrations add InitialCreate
Build started...
Build succeeded.
System.InvalidOperationException: The current CSharpHelper cannot scaffold literals of type 'System.Data.SqlTypes.SqlBytes'. Configure your services to use one that can.
   at Microsoft.EntityFrameworkCore.Design.Internal.CSharpHelper.UnknownLiteral(Object value)
   at Microsoft.EntityFrameworkCore.Design.Internal.CSharpHelper.Literal(Object[,] values)
   at Microsoft.EntityFrameworkCore.Migrations.Design.CSharpMigrationOperationGenerator.Generate(InsertDataOperation operation, IndentedStringBuilder builder)
   at System.Dynamic.UpdateDelegates.UpdateAndExecuteVoid3[T0,T1,T2](CallSite site, T0 arg0, T1 arg1, T2 arg2)
   at Microsoft.EntityFrameworkCore.Migrations.Design.CSharpMigrationOperationGenerator.Generate(String builderName, IReadOnlyList`1 operations, IndentedStringBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.Design.CSharpMigrationsGenerator.GenerateMigration(String migrationNamespace, String migrationName, IReadOnlyList`1 upOperations, IReadOnlyList`1 downOperations)
   at Microsoft.EntityFrameworkCore.Migrations.Design.MigrationsScaffolder.ScaffoldMigration(String migrationName, String rootNamespace, String subNamespace, String language)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
The current CSharpHelper cannot scaffold literals of type 'System.Data.SqlTypes.SqlBytes'. Configure your services to use one that can.
```

Is there something that needs adding so that EF Core will generate the migration?

**Comment by ajcvickers** [comment]
@chrisrichards This looks like it might be a bug--can you create a new issue?

**Comment by chrisrichards** [comment]
@ajvickers on the efcor repo?

**Comment by ajcvickers** [comment]
@chrisrichards Yes, thanks.

**Comment by chrisrichards** [comment]
@ajcvickers see #19785, thanks

**Comment by pantonis** [comment]
Is this supported in EF Core 3.1? If yes are there any examples that we can use?

**Comment by ajcvickers** [comment]
@pantonis Essentially, no. It might work for some scenarios on .NET Framework. It won't work on .NET Core since .NET Core does not have that type. See https://github.com/dotnet/SqlClient/issues/322

**Comment by pantonis** [comment]
:( Any plans to support it on .NET Core?

**Comment by bricelam** [comment]
@pantonis I don't know who has the best fork, but some clever folks [finished my initial experimentation](https://github.com/bricelam/EFCore.SqlServer.HierarchyId/network/members) of using the same approach we used to implement spatial.

**Comment by bricelam** [comment]
@aljones We should publish a package to NuGet.org

**Comment by pantonis** [comment]
@bricelam Still dont understand why EF6.3 which was supposed to not get any updates got this important feature and EF Core which is the new era didn't get it. :(

**Comment by bricelam** [comment]
@pantonis It was contributed to EF6--we just clicked merge. We probably would have done this on EF Core as part of the spatial implementation if Microsoft.SqlServer.Types was supported on .NET Core or even if there was a third-party library like NTS for hierarchyid. Basically, it doesn't seem prudent four our small team to implement, support, and maintain a HierarchyId library right now, so we're effectively blocked on implementing this.

**Comment by pantonis** [comment]
@bricelam I totally understand what you are saying. you guys are doing great job but please allow me to ask. Why is Microsoft not investing in EF Core by assigning more resources? I mean it is crazy. Lots of important functionality for an ORM is missing from EF Core (not saying about hierachyId). Is there any specific reason for very very very few resources on ef core development team?

**Comment by bricelam** [comment]
🤷🏻‍♂️ I just work here. And unfortunately, I'm not sure where the best place to provide that feedback is (which might be part of the problem).

**Comment by bricelam** [comment]
Some conversation happened here: https://developercommunity.visualstudio.com/idea/793848/increase-budget-for-entity-framework-core-dev-team.html

**Comment by bricelam** [comment]
But we're pretty off-topic. Let's take this conversation somewhere else. (Sorry everyone watching for hierarchyid)

**Comment by ErikEJ** [comment]
@ dotmorten - can your library help?

**Comment by bricelam** [comment]
@ErikEJ That's the approach used by [aljones/EFCore.SqlServer.HierarchyId](https://github.com/aljones/EFCore.SqlServer.HierarchyId) (the experiment I bootstrapped). We use Morten's implementation of Microsoft.SqlServer.Types for the hierarchyid serialization and logic.

**Comment by ErikEJ** [comment]
Is it published on nuget?

**Comment by bricelam** [comment]
Filed https://github.com/efcore/EFCore.SqlServer.HierarchyId/issues/1

**Comment by pantonis** [comment]
@bricelam Thanks a lot. IS this gonna be part of EF Core 3.1 update?

**Comment by bricelam** [comment]
To be clear, this is an unsupported third-party package. But yes, it will work with EF Core 3.1

**Comment by bricelam** [comment]
Published. See https://github.com/efcore/EFCore.SqlServer.HierarchyId

**Comment by StevenRasmussen** [comment]
@bricelam - Thanks!  I just tried installing and the package/s are installed fine but I'm getting some build errors: 

![image](https://user-images.githubusercontent.com/11198735/80153484-d8eaeb00-857a-11ea-8700-f4fb6fc3eab2.png)

I think there might be an issue with the way that the 'Abstraction' package is referenced from the main package.  It seems like it might be linked as a project reference instead of a package.  If I use one of the 'Quick Actions and Refactorings' provided by Visual Studio to try and add a reference it opens up a dialog to add a Project reference instead of NuGet package manager.  Anyway, happy to proved more details if necessary.  Thanks for your work on this!

**Comment by StevenRasmussen** [comment]
@bricelam - Sorry... false alarm!  I guess I hadn't removed all of my references properly from when I had the project locally.  After removing all references to the HierarchyId project/dll and using the NuGet packages fixes everything.  Thanks again!

**Comment by vyrotek** [comment]
Official support is sorely needed!

**Comment by Ogglas** [comment]
@bricelam @vyrotek Official support would still be appreciated :)

**Comment by bricelam** [comment]
For sure. We're still waiting on the SQL Server team to release a version of Microsoft.SqlServer.Types that works cross-platform on .NET 6+. Last I heard, they had a plan and were actively working on it.

**Comment by IT-CASADO** [comment]
Any news about official support?

**Comment by bricelam** [comment]
This is finally unblocked. Recent previews of the [Microsoft.SqlServer.Types](https://www.nuget.org/packages/Microsoft.SqlServer.Types) package support modern .NET and its hiererchyid type works cross-platform.

**Comment by mojtabakaviani** [comment]
Thanks everyone!  👍

**Comment by marchy** [comment]
What is the Sqlite support for this on EF8?

Curious if it's worthwhile updating our code to support the new EF8 hierarchy ID abstraction over simply doing string checks for ancestors on a hierarchical-by-convention ID (ie: `Ancestor1/Ancestor2/Node`) as we currently do.

**Comment by ajcvickers** [comment]
@marchy SQLite does not natively support HierarchyId, and we don't plan to add anything to fake it in EF Core.

**Comment by roji** [comment]
@marchy assuming you're using SQLite as a fake database for testing (with the real database being SQL Server), this is a good example of the limitations of that testing approach. We recommend using e.g. [testcontainers.net](https://testcontainers.com/guides/getting-started-with-testcontainers-for-dotnet/) to effortlessly spin up a real, containerized SQL Server database in your tests, and test against that.
